client/src/context/useNotificationContext.tsx

```diff
@@ -0,0 +1,94 @@
+import { useState, useContext, createContext, FunctionComponent, SyntheticEvent, useCallback, useEffect } from 'react';
+import { Notification } from '../interface/Notification';
+import { useAuth } from './useAuthContext';
+import { useSocket } from './useSocketContext';
+import getNotifications from '../helpers/APICalls/getNofitications';
+import postNotification from '../helpers/APICalls/postNotification';
+import patchNotifications from '../helpers/APICalls/patchNotifications';
+import {
+  NotificationApiData,
+  NotificationApiDataGet,
+  NotificationApiDataGetAll,
+  NotificationApiDataPost,
+} from '../interface/NotificationApiData';
+import { useSnackBar } from './useSnackbarContext';
+
+interface NotificationContext {
+  notifications: [Notification] | undefined;
+  pushNotification: (data: NotificationApiDataPost) => void;
+  readNotifications: () => void;
+}
+
+export const NotificationContext = createContext<NotificationContext>({
+  notifications: undefined,
+  pushNotification: () => null,
+  readNotifications: () => null,
+});
+
+export const NotificationContextProvider: FunctionComponent = ({ children }): JSX.Element => {
+  const [notifications, setNotifications] = useState<[Notification] | undefined>(undefined);
+  const { profile } = useAuth();
+  const { socket } = useSocket();
+  const { updateSnackBarMessage } = useSnackBar();
+
+  const pushNotification = useCallback(
+    async (data: NotificationApiDataPost) => {
+      await postNotification({
+        type: data.type,
+        description: data.description,
+        read: data.read,
+        createdBy: data.createdBy,
+        creatorName: data.creatorName,
+        creatorPhotoKey: data.creatorPhotoKey,
+        receivedBy: data.receivedBy,
+      }).then((data: NotificationApiData) => {
+        if (data.success) {
+          const { notification } = data.success as NotificationApiDataGet;
+          socket?.emit('notification', notification);
+        } else {
+          if (data.error) updateSnackBarMessage(data.error.message);
+        }
+      });
+    },
+    [socket, updateSnackBarMessage],
+  );
+
+  const readNotifications = useCallback(() => {
+    if (notifications && notifications.length >= 1) {
+      const newNotifications = [...notifications];
+      newNotifications.forEach((notification) => (notification.read = true));
+      setNotifications(newNotifications as any);
+      patchNotifications();
```

Comment:
We would only want to mark our notifications in our local state as read after this API call is successful :)

Reply:
changed so that notification in local state is only marked after patchNotifications return success.